### ISSUE TITLE
Question: Extending grammars and generating queries programatically

### ISSUE THREAD

**Issue by felipesere** [tree-sitter/tree-sitter#646]
I'm trying to write a linter/rule-enforcer for terraform files here https://github.com/felipesere/terracheck

My idea is that the tool would particularly useful if

* was built on something more powerful than a string of regexes and if-else chains

* the way to describe a "rule" would be intuitive if one already knew how to write Terraform code

So I thought I could parse Terraform with tree-sitter (I have a mostly working grammar already) and then turn the rules into queries as S-expressions to let tree-sitter do the heavy lifting of matching a rule against the AST.

I have a few questions though, as I think I might need more than one grammar:

Can I use a different (simpler) grammer to parse my rules and generate query S-expressions for the proper Terraform AST? My hunch is that the answer is 'no' because the node-types in the AST matter when matching queries. 

Alternatively, how could I go about extending my terraform grammar.js to support additional grammar-rules such as having `${*}` in pretty much all possible places to describe "Terraform, but with a rudimentary query langauge added on top" in a separate grammar? 

Would it be easier/smarter to just use a very rudimentary AST for terraform that in turn can be used for the rules? At the moment I have details such as `booleans`, `string_literals`, `ternary` in the proper Terraform-grammar  that I would have to give up to not have to introduce them in the rule-grammar...

Below is an example from my README that might help in clarifying.

## Example

As sample rule could be

> We only want PostgreSQL RDS instances

and then there could be a markdown document saying something like

```markdown
nr: 1235
---
# Only allow PostgreSQL RDS instances

Some clever rationale comparing the driver support to MySQL and Oracle


## Allow: Postgres RDS instance

resource "aws_db_instance" * {
  engine = "postgres"
}

## Deny - any other DB engine

resource "aws_db_instance" * {
}

...possibly more allow/deny blocks...

## Exceptions

* module.warehouse.aws_db_instance.main
* id-123u04r902380
* path/to/file.tf

```

We'd run throught the terraform code looking for matches resources that match
either `## Allow` or `## Deny` blocks and then flag as apropriate.

There would be some mechanism for describing placeholders and constraints on attributes names and connections.

For example, if we had the following terraform:

```terraform
resource "aws_db_instance" "postgresql" {
  allocated_storage               = var.allocated_storage
  backup_retention_period         = var.backup_retention_period
  db_subnet_group_name            = aws_db_subnet_group.main.name
  enabled_cloudwatch_logs_exports = ["postgresql"]
  
  engine                          = "postgres"
  engine_version                  = "9.6"
  
  final_snapshot_identifier       = "${var.identifier}-${var.environment}-final-db-snapshot"
  identifier                      = var.identifier
  instance_class                  = var.instance_class
  kms_key_id                      = aws_kms_key.db_encryption.arn
  multi_az                        = var.multi_az
  name                            = random_string.db_name.result
}

```

it would be nice if the following rule matched it:

```
resource "aws_db_instance" * {
  engine = "postgres"
}
```

The semantically equivalent query in `tree-sitter` looks like this:

```lisp
(
  (resource
    (resource_type) @type
    (block
      (attribute
        (identifier) @id
        (boolean) @val
  )))
  (#eq? @type "\"aws_db_instance\"")
  (#eq? @id "engine")
  (#eq? @val \"postgres\"")
) @result
```

The second "deny" rule 

```
resource "aws_db_instance" * {
}
```

would look like:

```lisp
(
 (resource
    (resource_type) @type)
  (#eq? @type "\"aws_db_instance\"")
)
```

**Comment by maxbrunsfeld** [comment]
> Can I use a different (simpler) grammer to parse my rules and generate query S-expressions for the proper Terraform AST?

Tree-sitter doesn't have functionality for generating queries based on source code, but I think if you wanted to write this functionality yourself, you could potentially do so using a simpler grammar.

> Alternatively, how could I go about extending my terraform grammar.js to support additional grammar-rules such as having ${*} in pretty much all possible places to describe "Terraform, but with a rudimentary query langauge added on top" in a separate grammar?

We haven't written docs on how to extend a grammar to create a derived grammar, but it's not too complex. You can look at the [C++](https://github.com/tree-sitter/tree-sitter-cpp/blob/5e7476bd014445abdae879661e9caf299215478a/grammar.js#L9) or [TypeScript](https://github.com/tree-sitter/tree-sitter-typescript/blob/ebd10b475722d59a1fa7e4b38047e768413794fc/common/define-grammar.js#L24) grammars for examples.

Basically, you pass the base grammar as the first argument to the `grammar` function. Then, when defining a rule, you can define it like [this](https://github.com/tree-sitter/tree-sitter-typescript/blob/ebd10b475722d59a1fa7e4b38047e768413794fc/common/define-grammar.js#L249-L260) (with a second parameter usually called `previous`), which gives you the grammar of the same name from the base grammar.

> Would it be easier/smarter to just use a very rudimentary AST for terraform that in turn can be used for the rules? 

I'm not sure; I'd have to work on the problem a bit. I could imagine this working, or you might be able to just use the same grammar.
